native/membrane_videocompositor/src/elixir_bridge/scene_validation.rs

```diff
@@ -0,0 +1,392 @@
+            Object::Texture(texture) => texture.mentioned_names(),
+            Object::Video(_) => Vec::new(),
+        }
+    }
+}
+
+impl Texture {
+    fn mentioned_names(&self) -> Vec<&Name> {
+        let mut names = vec![&self.input];
+
+        if let TextureOutputResolution::Name(name) = &self.resolution {
+            names.push(name);
+        }
+
+        names
+    }
+}
+
+impl Layout {
+    fn mentioned_names(&self) -> Vec<&Name> {
+        let mut names = self.inputs.values().collect::<Vec<_>>();
+
+        if let LayoutOutputResolution::Name(name) = &self.resolution {
+            names.push(name);
+        }
+
+        names
+    }
+}
+
+impl Scene {
+    fn check_for_duplicate_pad_refs(
+        objects: &[(ObjectName, Object)],
+    ) -> Result<(), SceneParsingError> {
+        let mut set = HashSet::new();
+
+        for (_, object) in objects {
+            if let Object::Video(InputVideo { input_pad }) = object {
+                if set.contains(input_pad) {
+                    return Err(SceneParsingError::DuplicatePadReferences(input_pad.clone()));
+                }
+
+                set.insert(input_pad);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn check_for_duplicate_or_undefined_names(
+        objects: &[(ObjectName, Object)],
+        final_object_name: &ObjectName,
+    ) -> Result<(), SceneParsingError> {
+        if objects.is_empty() {
+            return Err(SceneParsingError::UndefinedName(final_object_name.clone()));
+        }
+
+        let mut names = HashSet::new();
+        for (name, _) in objects {
+            if names.contains(name) {
+                return Err(SceneParsingError::DuplicateNames(name.clone()));
+            }
+
+            names.insert(name);
+        }
+
+        for (_, object) in objects {
+            for name in object.mentioned_names() {
+                if !names.contains(name) {
+                    return Err(SceneParsingError::UndefinedName(name.clone()));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn check_for_unused_objects(
+        objects: &[(ObjectName, Object)],
+        final_object_name: &ObjectName,
+    ) -> Result<(), SceneParsingError> {
+        let defined_names = objects.iter().map(|(name, _)| name).collect::<HashSet<_>>();
+
+        let used_names = objects
+            .iter()
+            .flat_map(|(_, object)| object.mentioned_names())
+            .collect::<HashSet<_>>();
+
+        let difference = defined_names.difference(&used_names).collect::<Vec<_>>();
+
+        if difference.is_empty() {
+            return Err(SceneParsingError::CycleDetected);
+        }
+
+        if *difference[0] != final_object_name {
+            return Err(SceneParsingError::UnusedObject((*difference[0]).clone()));
+        }
+
+        if difference.len() > 1 {
+            return Err(SceneParsingError::UnusedObject((*difference[1]).clone()));
+        }
+
+        Ok(())
+    }
```

Comment:
All names are used somehow. `final_object` is just referenced as `filan_object`. The naming convention here might be confusing.
```suggestion
    fn check_for_unused_objects(
        objects: &[(ObjectName, Object)],
        final_object_name: &ObjectName,
    ) -> Result<(), SceneParsingError> {
        let defined_names = objects.iter().map(|(name, _)| name).collect::<HashSet<_>>();

        let mut used_names = objects
            .iter()
            .flat_map(|(_, object)| object.mentioned_names())
            .collect::<HashSet<_>>();

        if used_names.contains(final_object_name) {
            return Err(SceneParsingError::CycleDetected);
        } else {
            used_names.insert(final_object_name);
        }

        let unused_names = defined_names.difference(&used_names).collect::<Vec<_>>();

        if !unused_names.is_empty() {
            return Err(SceneParsingError::UnusedObject((*unused_names[0]).clone()));
        }

        Ok(())
    }
```